### PR TITLE
Add optional playUntil method to Player interface

### PR DESCRIPTION
**Pull request by defunctzombie**
**User-Facing Changes**
Forward seek and backwards seek step forward and backwards the same amount of time.

**Description**
Our current behavior for forward seeking (via arrow key or seek button), is to "play" the data forward at least the seek amount. The purpose of this behavior is to ensure replay of stateful messages (like tf and markers) so pressing "seek" from the start of the file results in the same playback as playing from the start of the file.

Unfortunately, the way we implemented this feature was to start playback and attempt to pause playback after the seek target was reached. This behavior would result in a forward seek moving forward different and variable amounts of time since the "pause" would be a best effort invocation after some frames of messages.

This change removes this workaround play/quick-pause behavior in favor of first-class support within the player. The player can optionally implement a `playUntil` method. If available, seeking forward will invoke this method instead of regular `seek`. The player can implement `playUntil` however it wants. For IterablePlayer, the implementation emits all the messages in one "tick" to make forward seeking a small amount appear like a jump.

Fixes: #3980
Fixes: #3872
Fixes: #4059

### CODE REVIEW

**Review by jtbandes — COMMENTED** [review]
I'm not sure we should make this argument optional in the API – if the calling code is expecting the player to play until a specific time, it would be bad if the player just started playing and never paused

**Comment by defunctzombie** [comment]
> I'm not sure we should make this argument optional in the API – if the calling code is expecting the player to play until a specific time, it would be bad if the player just started playing and never paused

Yea - I think I landed here as well. My main hesitation was all the code I'd need to update 😭 Maybe its not that much or maybe I can wait until the great RAP purge.

I also thought about making it a separate player api method that could be optional. Maybe that is a nicer route to go. `startPlaybackUntil` ?

**Comment by defunctzombie** [comment]
@jtbandes I've re-worked this to add a new optional `playUntil` method on the player interface. After writing the description I'm wondering if it should be instead a `seekUntil`. I'll wait for you to review and give any thoughts on naming or implementation.

File: packages/studio-base/src/players/IterablePlayer/IterablePlayer.ts [hunk 1]
```diff
@@ -718,13 +732,12 @@ export class IterablePlayer implements Player {
       throw new Error("Invariant: Tried to play with no current time.");
     }
 
+    // If we have an until time, we will read all the messages until that time in one tick
+    const targetTime = this._untilTime ?? add(this._currentTime, fromMillis(rangeMillis));
+
```
Comment (jtbandes) [inline packages/studio-base/src/players/IterablePlayer/IterablePlayer.ts:737]: Are you assuming that the untilTime is not far in the future? That seems dangerous.

File: packages/studio-base/src/players/UserNodePlayer/index.ts [hunk 2]
```diff
@@ -915,16 +915,47 @@ export default class UserNodePlayer implements Player {
...
-  callService = async (service: string, request: unknown): Promise<unknown> =>
-    await this._player.callService(service, request);
-  startPlayback = (): void => this._player.startPlayback?.();
-  pausePlayback = (): void => this._player.pausePlayback?.();
-  setPlaybackSpeed = (speed: number): void => this._player.setPlaybackSpeed?.(speed);
-  seekPlayback = (time: Time, backfillDuration?: Time): void =>
-    this._player.seekPlayback?.(time, backfillDuration);
-  requestBackfill = (): void => this._player.requestBackfill();
+  setPublishers(publishers: AdvertiseOptions[]): void {
+    return this._player.setPublishers(publishers);
```
Comment (jtbandes) [inline packages/studio-base/src/players/UserNodePlayer/index.ts:919]: Remove all these `return`s, the method is `void`

File: packages/studio-base/src/players/types.ts [hunk 3]
```diff
@@ -65,6 +65,8 @@ export interface Player {
   startPlayback?(): void;
   pausePlayback?(): void;
   seekPlayback?(time: Time, backfillDuration?: Time): void;
+  // Advanced playback controls.
```
Comment (jtbandes) [inline packages/studio-base/src/players/types.ts:68]: This comment is not helping anyone

**Review by jtbandes — APPROVED** [review]
Looks ok, would avoid the assumption that untilTime is in the near future. This feature can eventually evolve into breakpoints: https://github.com/foxglove/studio/issues/2283

File: packages/studio-base/src/players/types.ts (hunk 3 above)
Comment (defunctzombie) [reply packages/studio-base/src/players/types.ts:68]: It's helping my line count 🤣

File: packages/studio-base/src/players/IterablePlayer/IterablePlayer.ts (hunk 1 above)
Comment (defunctzombie) [reply packages/studio-base/src/players/IterablePlayer/IterablePlayer.ts:737]: is there a reasonable assumption to clamp here? Maybe if <= 100ms then we can emit in one batch? Or maybe don't bother?

**Comment by defunctzombie** [comment]
Maybe I need to guard against untilTime being set to some time behind the current playhead?

**Comment by jtbandes** [comment]
That would also be a good thing to chec[](url)

**Pull request merged**